File: Circular_Doubly_Linked_List.java
import java.util.Scanner;
import java.io.IOException;

class Node {
    String data;
    Node next, prev;
    Node(String data) { this.data = data; }
}

class NewsTicker {
    Node head = null;
    int size = 0;

    void insertNews(String news) {
        Node newNode = new Node(news);
        if (head == null) {
            head = newNode;
            head.next = head.prev = head;
        } else {
            Node tail = head.prev;
            tail.next = newNode;
            newNode.prev = tail;
            newNode.next = head;
            head.prev = newNode;
        }
        size++;
        System.out.println("[Sistem] Berita berhasil ditambahkan.");
    }

    void deleteNews(int pos) {
        if (head == null || pos < 1 || pos > size) {
            System.out.println("[Sistem] Nomor urut tidak valid!");
            return;
        }
        Node curr = head;
        for (int i = 1; i < pos; i++) curr = curr.next;
        if (size == 1) head = null;
        else {
            curr.prev.next = curr.next;
            curr.next.prev = curr.prev;
            if (curr == head) head = curr.next;
        }
        size--;
        System.out.println("[Sistem] Berita berhasil dihapus.");
    }

    // FITUR UTAMA: Berhenti tanpa mematikan program
    void runTicker(boolean forward) {
        if (head == null) {
            System.out.println("[Sistem] Daftar berita kosong.");
            return;
        }
        
        System.out.println("\n=== SIMULASI TEXT BERJALAN ===");
        System.out.println(">>> Tekan ENTER untuk kembali ke menu utama <<<");
        
        Node curr = forward ? head : head.prev;
        
        try {
            // Selama tidak ada tombol yang ditekan (System.in.available == 0)
            while (System.in.available() == 0) {
                System.out.print("\r>>> NEWS: " + curr.data + " <<<          ");
                Thread.sleep(3000); 
                curr = forward ? curr.next : curr.prev;
            }
            // Membersihkan sisa input agar tidak mengganggu menu
            System.in.read(); 
        } catch (Exception e) {
            System.out.println("[Sistem] Terjadi interupsi.");
        }
        System.out.println("\n[Sistem] Berhenti. Kembali ke menu utama...");
    }

    void displaySpecific(int pos) {
        if (head == null || pos < 1 || pos > size) {
            System.out.println("[Sistem] Berita tidak ditemukan!");
            return;
        }
        Node curr = head;
        for (int i = 1; i < pos; i++) curr = curr.next;
        System.out.println("\n[Sistem] Berita No." + pos + ": " + curr.data);
    }
}

public class Circular_Doubly_Linked_List {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        NewsTicker nt = new NewsTicker();
        int pilihan = 0;

        while (pilihan != 6) {
            System.out.println("\n===== MENU NEWS TICKER =====");
            System.out.println("1. Insert Berita\n2. Hapus Berita\n3. Tampilkan Forward (Loop)\n4. Tampilkan Backward (Loop)\n5. Berita Spesifik\n6. Exit");
            System.out.print("Pilih Menu: ");
            try {
                pilihan = sc.nextInt();
                sc.nextLine(); 
                switch (pilihan) {
                    case 1: System.out.print("Teks: "); nt.insertNews(sc.nextLine()); break;
                    case 2: System.out.print("Hapus No: "); nt.deleteNews(sc.nextInt()); break;
                    case 3: nt.runTicker(true); break;
                    case 4: nt.runTicker(false); break;
                    case 5: System.out.print("No Berita: "); nt.displaySpecific(sc.nextInt()); break;
                }
            } catch (Exception e) {
                System.out.println("[Error] Input tidak valid.");
                sc.nextLine();
            }
        }
        sc.close();
    }
}

File: Circular_Doubly_Linked_List.py
import time
import sys

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class NewsTicker:
    def __init__(self):
        self.head = None
        self.size = 0

    # 1. Insert Berita (Ditambahkan di akhir)
    def insert_news(self, news):
        new_node = Node(news)
        if not self.head:
            self.head = new_node
            self.head.next = self.head
            self.head.prev = self.head
        else:
            tail = self.head.prev
            tail.next = new_node
            new_node.prev = tail
            new_node.next = self.head
            self.head.prev = new_node
        self.size += 1
        print("[Sistem] Berita berhasil ditambahkan!")

    # 2. Hapus Berita (Berdasarkan nomor urut)
    def delete_news(self, pos):
        if not self.head or pos < 1 or pos > self.size:
            print("[Sistem] Nomor urut tidak valid!")
            return
        
        curr = self.head
        for _ in range(pos - 1):
            curr = curr.next
        
        if self.size == 1:
            self.head = None
        else:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
            if curr == self.head:
                self.head = curr.next
        
        self.size -= 1
        print(f"[Sistem] Berita '{curr.data}' berhasil dihapus.")

    # 3 & 4. Tampil Berita (Horizontal & Infinite Loop)
    def run_ticker(self, forward=True):
        if not self.head:
            print("[Sistem] Belum ada berita.")
            return

        print("\n=== SIMULASI RUNNING TEXT (Tekan Ctrl+C untuk Berhenti) ===")
        curr = self.head if forward else self.head.prev
        
        try:
            while True:  # Looping selamanya
                # \r mengembalikan kursor ke awal baris agar teks tertimpa
                sys.stdout.write(f"\r>>> BERITA TV: {curr.data} <<<          ")
                sys.stdout.flush()
                time.sleep(3)
                curr = curr.next if forward else curr.prev
        except KeyboardInterrupt:
            print("\n\n[Sistem] Kembali ke menu...")

    # 5. Tampil Berita Tertentu (Berdasarkan nomor urut)
    def display_specific(self, pos):
        if not self.head or pos < 1 or pos > self.size:
            print("[Sistem] Berita tidak ditemukan!")
            return
        curr = self.head
        for _ in range(pos - 1):
            curr = curr.next
        print(f"\n[Sistem] Berita No.{pos}: {curr.data}")

def main():
    ticker = NewsTicker()
    while True:
        print("\n" + "="*25)
        print("   MENU NEWS TICKER")
        print("="*25)
        print("1. Insert Berita")
        print("2. Hapus Berita")
        print("3. Tampil Forward")
        print("4. Tampil Backward")
        print("5. Tampil Berita spesifik")
        print("6. Exit")
        
        pilih = input("Pilih Menu: ")
        
        if pilih == '1':
            ticker.insert_news(input("Teks berita: "))
        elif pilih == '2':
            try:
                ticker.delete_news(int(input("Nomor urut dihapus: ")))
            except ValueError: print("[Error] Masukkan angka!")
        elif pilih == '3':
            ticker.run_ticker(forward=True)
        elif pilih == '4':
            ticker.run_ticker(forward=False)
        elif pilih == '5':
            try:
                ticker.display_specific(int(input("Nomor urut berita: ")))
            except ValueError: print("[Error] Masukkan angka!")
        elif pilih == '6':
            print("Keluar...")
            break

if __name__ == "__main__":
    main()